### PR TITLE
System tray + Status listener

### PR DESCRIPTION
**Pull request by Vzor-**
This is a combination of the Dorkbox SystemTray implementation and the Printer Status listener.
Adds status page in sample.html
Adds native system tray support

There are a lot of erroneous commits and comments in here, I'm going to probably fix that and reissue this PR.

### CODE REVIEW

**Comment by lite1979** [comment]
Ubuntu 16.04, Oracle Java 8

I managed to get QZ-Tray to crash quite a few times by repeatedly clicking the listen buttons.

I have 32GB of memory, and Java only appeared to be around 900MB at the time of the crashes

Errors vary; see 1, 2, and 3 below:  Link to gist of backtrace at bottom of post.

1.   [TRACE] 2017-08-24 23:01:40,361 @ qz.printer.action.WebApp:?
	Waiting for JavaFX..
Exception in thread "JavaFX Application Thread" Exception in thread "JavaFX Application Thread"

2.   #
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fe27fbb4265, pid=3489, tid=0x00007fe27ceb4700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_144-b01) (build 1.8.0_144-b01)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.144-b01 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libcups.so.2+0x34265]  httpAddrFamily+0x5
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/lite/hs_err_pid3489.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Aborted (core dumped)

3.   [INFO] 2017-08-24 22:55:52,285 @ qz.common.TrayManager:?
	Allowed localhost to access connected printers
[thread 139822064322304 also had an error]
[thread 139821821196032 also had an error]
^CAborted (core dumped)


For the first error (1), I had to 'kill -9' the PID to get back to a clean terminal and system tray.  I think this may have something to do with JavaFX,/HTML printing, and not necessarily your PR.

To reproduce a crash like 2 or 3, simply click on the listen buttons a few times until you get a null connection error.

https://gist.github.com/lite1979/6db1cac7b7918c3ba81d55812fad9151

**Comment by Vzor-** [comment]
@lite1979 I was unable to reproduce this. Do you have your system set up with trusted certs so you don't have to accept the listen requests?

**Comment by lite1979** [comment]
No; I have the warnings for page load and signing at the moment. I'll install a keypair and retest in a few hours.

**Comment by Vzor-** [comment]
> I'll install a keypair and retest in a few hours.

That isn't necessary, I don't have them either. Hmm, maybe it is because I am using OpenJDK instead of Oracle Java 8. Also holy crap, 32GB of memory.

File: src/qz/printer/status/CupsUtils.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
+package qz.printer.status;
+
+import com.sun.jna.Pointer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sun.net.util.URLUtil;
```
Comment (tresf) [inline src/qz/printer/status/CupsUtils.java]: `sun.net.util.URLUtil` is deprecated.  Is it needed or is this leftover from something that was done previously?

**Comment by tresf** [comment]
@lite1979's right.  Not sure what combination does it, but clicking the four buttons in succession can crash it rather easily.

Ubuntu 14.04, Java 8u144 from WebUpd8 Team PPA.

In this case, the "Request Current Status" triggered the actual crash.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fea3d53f5db, pid=26728, tid=0x00007fea3ab37700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_144-b01) (build 1.8.0_144-b01)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.144-b01 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libcups.so.2+0x315db]  httpAddrConnect2+0xdb
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/ubuntu/hs_err_pid26728.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by tresf** [comment]
@Vzor- unrelated to printer status, but 14.04 is using the Java System tray instead of the native one.  Is this intentional?
screen shot 2017-08-26 at 3 28 52 am

**Comment by Vzor-** [comment]
@tresf There are some outstanding bugs with the tray implementation, this commit gets it back to gtk, though the 'open file location' and link to the website are currently broken.

File: src/qz/printer/status/CupsUtils.java (hunk 1 above)
Comment (Vzor-) [reply src/qz/printer/status/CupsUtils.java]: I will look into alternatives.

**Comment by Vzor-** [comment]
Switching to oracle jdk made it so I could reproduce the issue. Working on a fix

**Comment by tresf** [comment]
Please bump this PR to use SystemTray 3.4 per https://github.com/dorkbox/SystemTray/releases/tag/Release_3.4.

Also (sorry) some fast-forwards from 2.0 have put some files in conflict.

**Comment by tresf** [comment]
Superseded by #351.

**Pull request closed without merging**